doc/source/ray-core/user-spawn-processes.rst

```diff
@@ -0,0 +1,106 @@
+Lifetimes of a User-Spawn Process
+=================================
+
+To avoid leaking user-spawned processes, Ray provides mechanisms to kill all user-spawned processes when a worker that starts it exits. This feature prevents GPU memory leaks from child processes (e.g., torch).
+
+We have 2 environment variables to handle subprocess killing on core worker exit:
+
+- ``RAY_kill_child_processes_on_worker_exit`` (default ``true``): Only works on Linux. If true, core worker kills all subprocesses on exit. This won't work if the core worker crashed or was killed by a signal. If a process is created as a daemon, e.g. double forked, it will not be killed by this mechanism.
+
+- ``RAY_kill_child_processes_on_worker_exit_with_raylet_subreaper`` (default ``false``): Only works on Linux greater than or equal to 3.4. If true, Raylet kills any subprocesses that were spawned by the core worker after the core worker exits. This works even if the core worker crashed or was killed by a signal. Even if a process is created as a daemon, e.g. double forked, it will still be killed by this mechanism. The killing happens within 10 seconds after the core worker death.
+
+Both killings happen recursively, meaning if the core worker spawns user process A, then user process A spawns user process B, on core worker exit (or crash) both A and B are killed.
+
+On non-Linux platforms, user-spawned process is not controlled by Ray. The user is responsible for managing the lifetime of the child processes. If the parent Ray worker process dies, the child processes will continue to run.
+
+
```

Comment:
What we provide is more like a last resort. We should mention that ideally child processes should catch the death of the parent and self-exit so this feature doesn't need to be turned on.

Reply:
Since I reversed the default sigchld handler in core worker to reap-all-dead, users need to toggle it back to really do the signal handling. See the new `⚠️ Caution` section.

Reply:
Oh, I mean not the signal handle on the parent side but on the child side: https://stackoverflow.com/questions/284325/how-to-make-child-process-die-after-parent-exits.

Reply:
Understood. Added Note in the last section in the intro.